openff/toolkit/typing/engines/smirnoff/forcefield.py

```diff
@@ -1134,6 +1138,160 @@ def parse_smirnoff_from_source(self, source):
         msg += f"Parsing failed with the following error:\n{exception_msg}\n"
         raise IOError(msg)
 
+    def _terms(self):
+        """
+        Generate a list of active term keys. These keys are can be used to find the
+        specific value. The order the keys will be the same, and will only differ
+        if the order is manually changed, or the list of active terms is changed.
+
+        This should be public, but we make it private to hide from docs
+
+        Returns
+        -------
+        Generator[:class:`openff.toolkit.typing.engines.smirnoff.parameters._ParameterTermKey`]
+
+        """
+
+        for ph_name in self._parameter_handlers:
+            ph = self.get_parameter_handler(ph_name)
+            yield from zip(repeat(ph_name), ph._terms())
+
+    def _terms_set_active(self, terms=None):
+        """
+        Set which terms should be generated when queried.
+
+        This should be public, but we make it private to hide from docs
+
+        Parameters
+        ----------
+        keys : List[str] or None, default None
+            A list of keys that should be active. If keys is None, all terms are
+            activated. An empty list will deactivate all terms.
```

Comment:
Yes, exactly. To me, calling the method with no arguments would lead to "default" behavior and turn everything on. Conversely, one can communicate intentionally activating no terms by supplying a list of terms, but that list is empty. This is why `[]` would deactivate everything. Implementation-wise, this just means that each parameter will iterate through the supplied list, and find that none of its terms are there, so it doesn't activate any. The assumption here is that the call first assumes everything is inactive (unless None is supplied). Is there a way that is more intuitive? I tried to avoid adding an explicit `_terms_set_inactive` because I think that would lead something less clean and confusion when thinking about the interaction between the separate methods.

Reply:
Hmm, I can see your reasoning but I'm still getting hung up on the `None` case. I think the `terms=[]` case should deactivate everything, so I don't have issues with that behavior.


I imagine the method working as if I'm asking the object to "please set some terms to be active," to which it asks "okay, which terms do you want to be active?" If I say "None," to me, says that communicates that I want no terms to be active.

To turn everything off, I would rather require the use to be explicit and pass through all terms, so something like `force_field._terms_set_active(terms=force_field._get_loaded_terms)` or something that passes through _all_ terms. 

Two other ideas that I'm less in support of

* Have a separate `_terms_set_all_active` that handles just this case
* Allow `_terms_set_active(terms='all')` and handle that case inside the function

> I tried to avoid adding an explicit _terms_set_inactive because I think that would lead something less clean and confusion when thinking about the interaction between the separate methods.

I don't totally understand your point here - are you arguing that `force_field._terms_set_active(terms=[])` is an acceptable way to use this function, even though what it actually does it se things to be inactive, despite the function name? Because if so, I think that's fine.

Reply:
So I think your second bullet, using `"all"`, is essentially the same way I treat `None`. I just wanted the ability of calling with no args, `terms_set_active()`, to activate everything in the list. The other way to solve this is to call `_terms_get_loaded()` and supply that list instead. Would this be better? Then `terms` would be a (required) positional arg rather than a kwarg.

Reply:
I think we're on the same page - I see the value in being able to quickly set everything to be active, but I also think `._terms_set_active()` is too ambiguous. Your ideas for behavior seem good and I think we're just working through the semantics - I was confused at first, so I would be worried that users could be as well.

> The other way to solve this is to call _terms_get_loaded() and supply that list instead. Would this be better?

Yes, I would personally prefer this approach. Your proposal to make `terms` a required positional argument would fix the ambiguity I mentioned above.